### ISSUE TITLE
32-bit builds of v0.13.8 panic (386, arm)

### ISSUE THREAD

**Issue by xduugu** [syncthing/syncthing#3347]
To fix: `syncthing -upgrade` and get the v0.13.9 release.

//@calmh

---

The latest linux arm release (0.13.8) panics for me. Removing the index does not help. Reverting to 0.13.7 fixes the issue.

Panic log:
```
[XXXXX] 12:36:28 INFO: syncthing v0.13.8 "Copper Cockroach" (go1.6.2 linux-arm) unknown-user@lx-buildslave.syncthing.net 2016-06-26 10:48:18 UTC
...
Panic at 2016-06-26T12:36:28Z
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x4 pc=0x366bc8]

goroutine 1 [running]:
panic(0x62a920, 0x10d0a038)
        /usr/local/go/src/runtime/panic.go:481 +0x330
sync/atomic.addUint64(0x1110e30c, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/sync/atomic/64bit_arm.go:31 +0x68
github.com/syncthing/syncthing/lib/db.readWriteTransaction.flush(0x10d136e0, 0x1110e300, 0x1110ef40)
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/lib/db/leveldb_transactions.go:74 +0xbc
github.com/syncthing/syncthing/lib/db.readWriteTransaction.close(0x10d136e0, 0x1110e300, 0x1110ef40)
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/lib/db/leveldb_transactions.go:59 +0x2c
github.com/syncthing/syncthing/lib/db.(*Instance).checkGlobals(0x1110e300, 0x111128c0, 0xd, 0x10, 0x10d524e4)
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/lib/db/leveldb_dbinstance.go:643 +0x870
github.com/syncthing/syncthing/lib/db.NewFileSet(0x11109fd0, 0xd, 0x1110e300, 0x10e6b3d4)
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/lib/db/set.go:104 +0x268
github.com/syncthing/syncthing/lib/model.(*Model).AddFolder(0x10d0c2c0, 0x11109fd0, 0xd, 0x0, 0x0, 0x1109bb00, 0x22, 0x0, 0x11081980, 0x3, ...)
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/lib/model/model.go:1378 +0xf8
main.syncthingMain(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6a4088, 0x1, 0x0, ...)
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:688 +0x18f8
main.main()
        /home/buildslave/workspace/syncthing-release/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:364 +0x748
```

**Comment by AndreB82** [comment]
Same error on Win x86 (Vista)

**Comment by xduugu** [comment]
Linux amd64 works fine. Maybe this issue affects all 32-bit builds.

**Comment by xduugu** [comment]
The linux i386 panics as well. I'll change the title of the issue.

**Comment by andede** [comment]
Same error for me on linux armv71(odroid XU4) and armv61(RPI 1b):

> [KWXTT] INFO: syncthing v0.13.8 "Copper Cockroach" (go1.6.2 linux-arm) unknown-user@lx-buildslave.syncthing.net 2016-06-26 10:48:18 UTC
Jun 26 14:31:13 EDLSF syncthing[21364]: [KWXTT] INFO: My ID: KWXTTWN-TEPOF4B-IHPFKVY-357NLHH-ZVI4JS2-ML4CL5C-JE4MAXL-LYDRHAO
Jun 26 14:31:13 EDLSF syncthing[21364]: [KWXTT] INFO: Single thread hash performance is ~34 MB/s
Jun 26 14:31:13 EDLSF syncthing[21364]: [KWXTT] INFO: Local networks: 127.0.0.1/8, 192.168.1.101/24, ::1/128, fe80::21e:6ff:fe31:13f0/64
Jun 26 14:31:15 EDLSF syncthing[21364]: panic: runtime error: invalid memory address or nil pointer dereference
Jun 26 14:31:15 EDLSF syncthing[21364]: [signal 0xb code=0x1 addr=0x4 pc=0x363014]
Jun 26 14:31:15 EDLSF syncthing[21364]: goroutine 1 [running]:
Jun 26 14:31:15 EDLSF syncthing[21364]: panic(0x613780, 0x10d20030)
Jun 26 14:31:15 EDLSF syncthing[21364]: /usr/local/go/src/runtime/panic.go:481 +0x330
Jun 26 14:31:15 EDLSF syncthing[21364]: sync/atomic.addUint64(0x1115a7ac, 0x0, 0x0, 0x0, 0x0)
Jun 26 14:31:15 EDLSF syncthing[21364]: /usr/local/go/src/sync/atomic/64bit_arm.go:31 +0x68
Jun 26 14:31:15 EDLSF syncthing[21364]: github.com/syncthing/syncthing/lib/db.readWriteTransaction.flush(0x121eb410, 0x1115a7a0, 0x121eee40)
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/lib/db/leveldb_transactions.go:74 +0xbc
Jun 26 14:31:15 EDLSF syncthing[21364]: github.com/syncthing/syncthing/lib/db.readWriteTransaction.close(0x121eb410, 0x1115a7a0, 0x121eee40)
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/lib/db/leveldb_transactions.go:59 +0x2c
Jun 26 14:31:15 EDLSF syncthing[21364]: github.com/syncthing/syncthing/lib/db.(*Instance).checkGlobals(0x1115a7a0, 0x121eee00, 0x11, 0x20, 0x121d8bc4)
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/lib/db/leveldb_dbinstance.go:643 +0x870
Jun 26 14:31:15 EDLSF syncthing[21364]: github.com/syncthing/syncthing/lib/db.NewFileSet(0x110dfc60, 0x11, 0x1115a7a0, 0x10e893dc)
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/lib/db/set.go:104 +0x268
Jun 26 14:31:15 EDLSF syncthing[21364]: github.com/syncthing/syncthing/lib/model.(*Model).AddFolder(0x10d88000, 0x110dfc60, 0x11, 0x0, 0x0, 0x1115bf80, 0x12, 0x0, 0x110d3880, 0x2, ...)
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/lib/model/model.go:1378 +0xf8
Jun 26 14:31:15 EDLSF syncthing[21364]: main.syncthingMain(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x68b6dc, 0x1, 0x0, ...)
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:688 +0x18f8
Jun 26 14:31:15 EDLSF syncthing[21364]: main.main()
Jun 26 14:31:15 EDLSF syncthing[21364]: /home/buildslave/workspace/syncthing-release-debian/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:364 +0x748
Jun 26 14:31:15 EDLSF systemd[1]: syncthing@root.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 26 14:31:15 EDLSF systemd[1]: syncthing@root.service: Unit entered failed state.
Jun 26 14:31:15 EDLSF systemd[1]: syncthing@root.service: Failed with result 'exit-code'.

Revert to 0.13.7 solve

**Comment by xduugu** [comment]
I guess it is because of commit b779e22205585ca1a05dc2e87adc82ec3b47d24a. The fix is the same as 707837b2ebe51783beee7b0594b1b5ea9b2f62f5.

**Comment by xduugu** [comment]
I've created a pull request #3350. The builds at https://build.syncthing.net/job/syncthing-pr/2516/artifact/ fix the issue for me on linux arm and i386.

**Comment by calmh** [comment]
Reopening for visibility

**Comment by calmh** [comment]
Resolved in v0.13.9 that I'm pushing now.

**Comment by AudriusButkevicius** [comment]
Surely a dup?
